sale.py

```diff
@@ -226,3 +245,120 @@ def process(cls, sales):
             if sale.state not in ('confirmed', 'processing', 'done'):
                 continue        # pragma: no cover
             sale.create_gift_cards()
+
+
+class Payment:
+    'Payment'
+    __name__ = 'sale.payment'
+
+    gift_card = fields.Many2One(
+        "gift_card.gift_card", "Gift Card", states={
+            'required': Eval('method') == 'gift_card',
+            'invisible': ~(Eval('method') == 'gift_card'),
+        }, domain=[('state', '=', 'active')], depends=['method']
+    )
+
+    def _create_payment_transaction(self, amount, description):
```

Comment:
This looks wrong. If the payment gateway's method design has to change, it has to, but overwriting so much is not a good idea.